ndonnx/_typed_array/onnx.py

```diff
@@ -0,0 +1,2125 @@
+
+
+class Integer(_Number[TY_ARRAY_co]): ...
+
+
+class Floating(_Number[TY_ARRAY_co]): ...
+
+
+class Int8(Integer["TyArrayInt8"]):
+    def _build(self, var: Var) -> TyArrayInt8:
+        return TyArrayInt8(var)
+
+
+class Int16(Integer["TyArrayInt16"]):
+    def _build(self, var: Var) -> TyArrayInt16:
+        return TyArrayInt16(var)
+
+
+class Int32(Integer["TyArrayInt32"]):
+    def _build(self, var: Var) -> TyArrayInt32:
+        return TyArrayInt32(var)
+
+
+class Int64(Integer["TyArrayInt64"]):
+    def _build(self, var: Var) -> TyArrayInt64:
+        return TyArrayInt64(var)
+
+
+class UInt8(Integer["TyArrayUInt8"]):
+    def _build(self, var: Var) -> TyArrayUInt8:
+        return TyArrayUInt8(var)
+
+
+class UInt16(Integer["TyArrayUInt16"]):
+    def _build(self, var: Var) -> TyArrayUInt16:
+        return TyArrayUInt16(var)
+
+
+class UInt32(Integer["TyArrayUInt32"]):
+    def _build(self, var: Var) -> TyArrayUInt32:
+        return TyArrayUInt32(var)
+
+
+class UInt64(Integer["TyArrayUInt64"]):
+    def _build(self, var: Var) -> TyArrayUInt64:
+        return TyArrayUInt64(var)
+
+
+class Float16(Floating["TyArrayFloat16"]):
+    def _build(self, var: Var) -> TyArrayFloat16:
+        return TyArrayFloat16(var)
+
+
+class Float32(Floating["TyArrayFloat32"]):
+    def _build(self, var: Var) -> TyArrayFloat32:
+        return TyArrayFloat32(var)
+
+
+class Float64(Floating["TyArrayFloat64"]):
+    def _build(self, var: Var) -> TyArrayFloat64:
+        return TyArrayFloat64(var)
+
+
+# Non-nullable Singleton instances
+bool_: Bool = Bool()
+
+float16: Float16 = Float16()
+float32: Float32 = Float32()
+float64: Float64 = Float64()
+
+int16: Int16 = Int16()
+int32: Int32 = Int32()
+int64: Int64 = Int64()
+int8: Int8 = Int8()
+
+uint8: UInt8 = UInt8()
+uint16: UInt16 = UInt16()
+uint32: UInt32 = UInt32()
+uint64: UInt64 = UInt64()
+
+utf8: Utf8 = Utf8()
+
+# Union types
+#
+# Union types are exhaustive and don't create ambiguities with respect to user-defined subtypes.
+FloatingDTypes = Float16 | Float32 | Float64
+SignedIntegerDTypes = Int8 | Int16 | Int32 | Int64
+UnsignedIntegerDTypes = UInt8 | UInt16 | UInt32 | UInt64
+IntegerDTypes = SignedIntegerDTypes | UnsignedIntegerDTypes
+NumericDTypes = FloatingDTypes | IntegerDTypes
+DTypes = NumericDTypes | Utf8 | Bool
+
+
+class TyArray(TyArrayBase):
+    _var: Var
+
+    def __init__(self, var: Var):
+        var_dtype = var.unwrap_tensor().dtype
+        if from_numpy(var_dtype) != self.dtype:
+            raise ValueError(
+                f"data type of 'var' (`{var_dtype}`) is incompatible with `{type(self)}`"
+            )
+        if var.unwrap_tensor().shape is None:
+            raise ValueError("'var' has no shape information")
+        self._var = var
+
+    def __ndx_value_repr__(self) -> dict[str, str]:
+        try:
+            # TODO: should this really be "data"?
+            return {"data": str(self.unwrap_numpy().tolist())}
+        except ValueError:
```

Comment:
This try-catch looks like it'll be buggy. Specifically, a `ValueError` inside `unwrap_numpy()` can also be from invalid data, _not just because there's no data_. In this file you can see this as an example: https://github.com/Quantco/ndonnx/pull/110/files#diff-7e2260bdac28e59ef26d7f8e6f2f350474002467a5107dbcaa6ecf4d091ac12cR585. I think you should expose an method returning a bool for if lazy or not explicitly.

Reply:
I see that it says it's unreachable in that case, but in principle it is user definable code and `ValueError` can emerge from anywhere so it makes little sense to assume it's because the array is lazy.

Reply:
I'm afraid the link is broken, but this is now done using the new `is_constant` property.